test/redis/interval_cache_test.exs

```diff
@@ -1,192 +1,92 @@
-  end
-
-  test "handles blank responses", %{from: from, to: to} do
-    {data, meta} = interval @prefix, from, to, QuarterHourly, fn(new_from) ->
-      assert format_dtim(new_from) == "2017-03-22T01:15:00Z"
-      {[], %{meta: "data"}}
-    end
-    assert redis_count("#{@prefix}*") == 6
-    assert Enum.uniq(Enum.map(data, &(&1.count))) == [0]
-    assert meta.meta == "data"
+  test "misses the whole interval", %{interval: intv} do
+    {data, meta} = interval(@prefix, intv, "foobar", &test_work_fn/1)
+    assert redis_count("#{@prefix}*") == 4
     assert meta.cache_hits == 0
+    assert length(data) == 4
+    assert_time data, 0, "2017-03-22T01:15:00Z"
+    assert_time data, 1, "2017-03-22T01:30:00Z"
+    assert_time data, 2, "2017-03-22T01:45:00Z"
+    assert_time data, 3, "2017-03-22T02:00:00Z"
+    assert Enum.at(data, 0).count == 13
+    assert Enum.at(data, 1).count == 0
+    assert Enum.at(data, 2).count == 0
+    assert Enum.at(data, 3).count == 43
+
+    {data2, meta2} = interval(@prefix, intv, "foobar", &test_work_fn/1)
+    assert redis_count("#{@prefix}*") == 4
+    assert meta2.cache_hits == 4
+    assert meta2.cached == true
+    assert data2 == data
   end
 
-  test "caches response intervals", %{from: from, partial_to: partial_to, to: to} do
-    data1 = [
-      %{count: 55, time: get_dtim("2017-03-22T01:15:00Z")},
-      %{count: 66, time: get_dtim("2017-03-22T01:30:00Z")},
-      %{count: 77, time: get_dtim("2017-03-22T01:45:00Z")},
-    ]
-    data2 = [
-      %{count: 88, time: get_dtim("2017-03-22T02:00:00Z")},
-      %{count: 99, time: get_dtim("2017-03-22T02:15:00Z")},
-      %{count: 111, time: get_dtim("2017-03-22T02:30:00Z")},
-    ]
-
-    {data, meta} = interval @prefix, from, partial_to, QuarterHourly, fn(new_from) ->
-      assert format_dtim(new_from) == "2017-03-22T01:15:00Z"
-      {data1, %{meta: "data"}}
-    end
-    assert redis_count("#{@prefix}*") == 3
-    assert Enum.map(data, &(&1.count)) == [55, 66, 77]
-    assert format_dtim(hd(data).time) == "2017-03-22T01:15:00Z"
-    assert meta.meta == "data"
-    assert meta.cache_hits == 0
-
-    {data, meta} = interval @prefix, from, to, QuarterHourly, fn(new_from) ->
-      assert format_dtim(new_from) == "2017-03-22T02:00:00Z"
-      {data2, %{meta: "stuff"}}
-    end
-    assert redis_count("#{@prefix}*") == 6
-    assert Enum.map(data, &(&1.count)) == [55, 66, 77, 88, 99, 111]
-    assert format_dtim(hd(data).time) == "2017-03-22T01:15:00Z"
-    assert meta.meta == "stuff"
-    assert meta.cache_hits == 3
-
-    {data, meta} = interval @prefix, from, to, QuarterHourly, fn(_new_from) ->
-      raise "should not have called this"
-    end
-    assert Enum.map(data, &(&1.count)) == [55, 66, 77, 88, 99, 111]
-    assert format_dtim(hd(data).time) == "2017-03-22T01:15:00Z"
-    assert meta.cached == true
-    assert meta.cache_hits == 6
+  test "partially hits the cache", %{interval: intv, keys: keys} do
```

Comment:
a little puzzled by this test -- it's testing that the attempt to set the same key (key #3) with a different value (44 instead of 43) won't work, right? i'm not sure i understand why `meta.cache_hits` is 2 and not 3 though, if 0, 1, and 3 are all set? i think i'm missing something...

Reply:
Since key-2 is a miss, it doesn't even check the cache for key-3.  It asks BQ for both 2 and 3, then assembles the result `[0-cached, 1-cached, 2-from-BQ, 3-from-BQ]`.

Reply:
ah, because key-2 would be present if it had been queried before, even if everything in it was 0. so if it misses one key it assumes nothing after it is present either. 👍